crates/rattler_networking/src/authentication_storage/netrc_storage.rs

```diff
@@ -0,0 +1,89 @@
+//! Fallback storage for passwords.
+use netrc_rs::{Machine, Netrc};
+use serde::de::value;
+use std::{
+    collections::HashMap,
+    env,
+    io::Read,
+    path::PathBuf,
+    sync::{Arc, Mutex},
+};
+
+/// A struct that implements storage and access of authentication
+/// information backed by a on-disk JSON file
+#[derive(Clone)]
+pub struct NetRcStorage {
+    /// The path to the JSON file
+    pub path: PathBuf,
+
+    /// A mutex to ensure that only one thread accesses the file at a time
+    mutex: Arc<Mutex<()>>,
+
+    /// The netrc file contents
+    machines: HashMap<String, Machine>,
+}
+
+/// An error that can occur when accessing the fallback storage
+#[derive(thiserror::Error, Debug)]
+pub enum NetRcStorageError {
+    /// An IO error occurred when accessing the fallback storage
+    #[error("IO error: {0}")]
+    IOError(#[from] std::io::Error),
+    // /// An error occurred when (de)serializing the credentials
+    // #[error("JSON error: {0}")]
+    // JSONError(#[from] serde_json::Error),
+}
+
+impl NetRcStorage {
+    /// Create a new fallback storage with the given path
+    pub fn new() -> Self {
```

Comment:
I would prefer to call this from_env since it reads the config from the environment. This means the environment has side effects when. I think this should also be reflected in a comment.